### PR TITLE
Indexing components and contours

### LINKED ISSUE #1117: Need to be able to see the component index
Relates to #1116.

### LINKED ISSUE #1157: Need to be able to see the contour index
Relates to #1117.

### PR DESCRIPTION
**Pull request by ollimeier**
Fixes #1117 and fixes #1157.

Reordering like requested in #1116 is a different task.

### CODE REVIEW

**Comment by justvanrossum** [comment]
Reordering is a very different task, that’s why it has a separate issue.

**Comment by ollimeier** [comment]
@justvanrossum I made two separate 'Show's: one for Contours, one for Components – but I was wondering if it would make more sense to bundle both into one 'Show shape index'? What do you think? Better keep them separated or have them bundled?

**Comment by justvanrossum** [comment]
> Better keep them separated or have them bundled?

Please keep then separate.

File: src/fontra/views/editor/visualization-layer-definitions.js [hunk 1]
```diff
@@ -543,6 +543,119 @@ registerVisualizationLayerDefinition({
   },
 });
 
+registerVisualizationLayerDefinition({
+  identifier: "fontra.contourIndex.indicator",
```
Comment (justvanrossum) [inline src/fontra/views/editor/visualization-layer-definitions.js:547]:
```suggestion
  identifier: "fontra.contour.index",
```

File: src/fontra/views/editor/visualization-layer-definitions.js [hunk 2]
```diff
@@ -543,6 +543,119 @@ registerVisualizationLayerDefinition({
...
+      context.fillStyle = parameters.color;
+      context.fillText(strLine, startPoint.x, -startPoint.y - bottomY);
+      startPointIndex = contourInfo.endPoint + 1;
+      contourIndex += 1;
+    }
+  },
+});
+
+registerVisualizationLayerDefinition({
+  identifier: "fontra.componentIndex.indicator",
```
Comment (justvanrossum) [inline src/fontra/views/editor/visualization-layer-definitions.js:597]:
```suggestion
  identifier: "fontra.component.index",
```

File: src/fontra/views/editor/visualization-layer-definitions.js [hunk 3]
```diff
@@ -543,6 +543,119 @@ registerVisualizationLayerDefinition({
...
+    const bottomY = -boxHeight / 2;
+
+    context.font = `${fontSize}px fontra-ui-regular, sans-serif`;
+    context.textAlign = "center";
+    context.scale(1, -1);
+
+    let shapeIndex = 0;
+    for (const componentController of glyph.components) {
+      const bounds = componentController._controlBounds;
+      let pt = { x: 0, y: 0 };
```
Comment (justvanrossum) [inline src/fontra/views/editor/visualization-layer-definitions.js:622]: Suggest to use this style instead (I bet prettier will wrap it better):
```js
    const pt = bounds ?
      {x: (bounds.xMax - bounds.xMin) / 2 + bounds.xMin, y: (bounds.yMax - bounds.yMin) / 2 + bounds.yMin} :
      {x: transformation.translateX, y: transformation.translateY};
```

**Review by justvanrossum — COMMENTED** [review]
Mostly looks good to me.

I need to look deeper into the bounds-not-yet-available problem: it's easiest to see when you drag a component (the label will flicker between origin and center, mostly at origin).

**Comment by justvanrossum** [comment]
Hi Olli, thanks for the good work!

I did some touch ups, and fixed the flickering/no-bounds problem (by using `compo.controlBounds` instead of `compo._controlBounds`: only ever use underscored properties if you know exactly what they are, `_controlBounds` is an internal property that client code shouldn't use).

I'm restarting our online deployments tonight (when it's quiet so I don't disturb designers), and I wanted this update to be included, that's why I did the small fixes instead of posting them as feedback.

**Pull request merged**